solana-programs/anchor/audius-data/programs/audius-data/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+            audius_user_acct.authority = user_authority;
+        }
+        Ok(())
+    }
+
+    /// Permissioned function to log an update to User metadata
+    pub fn update_user(ctx: Context<UpdateUser>, metadata: String) -> ProgramResult {
+        msg!("Audius::UpdateUser");
+        if ctx.accounts.user.authority != ctx.accounts.user_authority.key() {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+        msg!("AudiusUserMetadata = {:?}", metadata);
+        Ok(())
+    }
+
+    // User TODOS:
+    // - Disable audius admin signing
+    // - Enable happy path flow with both eth address and sol key
+}
+
+// Instructions
+#[derive(Accounts)]
+/// Instruction container to initialize an instance of Audius, with the incoming admin keypair
+pub struct Initialize<'info> {
+    #[account(init, payer = payer, space = 8 + 32)]
+    pub admin: Account<'info, AudiusAdmin>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+/// Instruction container to initialize a user account, must be invoked from an existing Audius
+/// `admin` account.
+/// `user` is a PDA derived from the Audius account and handle.
+/// `authority` is a signer key matching the admin value stored in AudiusAdmin root. Only the
+///  admin of this Audius root program may initialize users through this function
+/// `payer` is the account responsible for the lamports required to allocate this account.
+/// `system_program` is required for PDA derivation.
+#[derive(Accounts)]
+#[instruction(base: Pubkey, eth_address: [u8;20], handle_seed: [u8;16], user_bump: u8)]
+pub struct InitializeUser<'info> {
+    pub admin: Account<'info, AudiusAdmin>,
+    #[account(
+        init,
+        payer = payer,
+        seeds = [&base.to_bytes()[..32], handle_seed.as_ref()],
+        bump = user_bump,
+        space = 8 + 32 + 20
+    )]
+    pub user: Account<'info, User>,
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+/// Instruction container to allow a user to add their Solana public key as part of their identity.
+/// `user` is the target user PDA.
+/// The global sys var program is required to enable instruction introspection.
+#[derive(Accounts)]
+pub struct InitializeUserSolIdentity<'info> {
+    #[account(mut)]
+    pub user: Account<'info, User>,
+    pub sysvar_program: AccountInfo<'info>
+}
+
+/// Instruction container to allow updates to a given User account.
+/// `user` is the target user PDA.
+/// `user_authority` is a signer field which must match the `authority` field in the User account.
+#[derive(Accounts)]
+pub struct UpdateUser<'info> {
+    #[account(mut)]
+    pub user: Account<'info, User>,
+    #[account(mut)]
+    pub user_authority: Signer<'info>,
+}
+
+// END Instructions
+
+/// Audius root account
+#[account]
+pub struct AudiusAdmin {
```

Comment:
should we define these at the top since they're using the instructions?

Reply:
so this is me just following the prescriptive anchor layouts for now - i do think its easier to flow down from instructions -> state -> errors but happy to discuss